File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/db/result/entity/CommonWorkResult.kt
package jp.co.toukei.log.trustar.db.result.entity

import androidx.room.ColumnInfo
import androidx.room.Embedded
import androidx.room.Entity
import jp.co.toukei.log.lib.util.RequestBodyJson
import jp.co.toukei.log.trustar.db.user.embedded.LocationRecord
import jp.co.toukei.log.trustar.db.user.entity.BinDetail
import jp.co.toukei.log.trustar.user.UserInfo
import org.json.JSONObject
import third.jsonObj

@Entity(
    tableName = "common_work_result",
    primaryKeys = ["company_cd", "user_id", "allocation_no", "allocation_row_no"]
)
class CommonWorkResult(
    companyCd: String,
    userId: String,
    @ColumnInfo(name = "allocation_no") @JvmField val allocationNo: String,
    @ColumnInfo(name = "allocation_row_no") @JvmField val allocationRowNo: Int,

    @ColumnInfo(name = "work_cd") @JvmField val workCd: String?,
    @ColumnInfo(name = "status") @JvmField val status: Int,
    @ColumnInfo(name = "delay_status") @JvmField val delayStatus: Int,
    @ColumnInfo(name = "delay_reason_cd") @JvmField val delayReasonCd: String?,
    @ColumnInfo(name = "misdelivery_status") @JvmField val misdeliveryStatus: Int,
    @ColumnInfo(name = "org_allocation_row_no") @JvmField val orgAllocationRowNo: Int?,

    @Embedded(prefix = "start_") @JvmField val workStart: LocationRecord?,
    @Embedded(prefix = "end_") @JvmField val workEnd: LocationRecord?,
    @ColumnInfo(name = "temperature") @JvmField val temperature: Double?,
    @ColumnInfo(name = "experience_place_note1") @JvmField val experiencePlaceNote1: String?,
    @ColumnInfo(name = "operating_mode") @JvmField val operatingMode: Boolean,
) : RequestBodyJson, CommonUserSync(companyCd, userId) {

    constructor(userInfo: UserInfo, binDetail: BinDetail) : this(
        userInfo.companyCd,
        userInfo.userId,
        binDetail.allocationNo,
        binDetail.allocationRowNo,

        binDetail.work?.workCd,
        binDetail.status,
        binDetail.delayStatus,
        binDetail.delayReasonCd,
        binDetail.misdeliveryStatus,
        binDetail.origAllocationRowNo,

        binDetail.workStart,
        binDetail.workEnd,
        binDetail.temperature,
        binDetail.experiencePlaceNote1,
        binDetail.autoModeFlag
    )

    override fun jsonBody(): JSONObject {
        return jsonObj {
            "companyCd" v companyCd
            "userId" v userId
            "allocationNo" v allocationNo
            "allocationRowNo" v allocationRowNo

            "workCd" v workCd
            "status" v status
            "delayStatus" v (if (delayStatus == 0) 0 else 1)
            "delayReasonCd" v delayReasonCd
            "misdeliveryStatus" v misdeliveryStatus
            "orgAllocationRowNo" v orgAllocationRowNo

            workStart?.apply {
                "latitudeFrom" v (latitude ?: 0.0)
                "longitudeFrom" v (longitude ?: 0.0)
                "accuracyFrom" v (accuracy ?: 0F)
                "workDatetimeFrom" v date
            }
            workEnd?.apply {
                "latitudeTo" v (latitude ?: 0.0)
                "longitudeTo" v (longitude ?: 0.0)
                "accuracyTo" v (accuracy ?: 0F)
                "workDatetimeTo" v date
            }
            "temperature" v (temperature ?: JSONObject.NULL)
            "experiencePlaceNote1" v (experiencePlaceNote1 ?: JSONObject.NULL)

            if (operatingMode)
                "operatingMode" v "autoMode"
        }
    }
}

File: LogiOS/TrustarApp/Database/result/Entity/CommonWorkResult.swift
//
//  CommonWorkResult.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/01/31.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import GRDB

enum CommonWorkResultColumns: String, ColumnExpression {
    case companyCd, userId, allocationNo, allocationRowNo, workCd, status, delayStatus, delayReasonCd, misdeliveryStatus, orgAllocationRowNo
    case latitudeFrom, longitudeFrom, accuracyFrom, workDatetimeFrom
    case latitudeTo, longitudeTo, accuracyTo, workDatetimeTo
    case temperature, experiencePlaceNote1
    case operatingMode
}

class CommonWorkResult: BaseEntity, Codable {
    var companyCd: String             = Resources.strEmpty
    var userId: String                = Resources.strEmpty
    var allocationNo: String          = Resources.strEmpty
    var allocationRowNo: Int          = Resources.zeroNumber
    var workCd: String?               = nil
    var status: Int                   = Resources.zeroNumber
    var delayStatus: Int              = Resources.zeroNumber
    var delayReasonCd: String?        = nil
    var misdeliveryStatus: String     = Resources.strEmpty
    var orgAllocationRowNo: Int?      = Resources.zeroNumber

    var latitudeFrom: Double?        = nil
    var longitudeFrom: Double?       = nil
    var accuracyFrom: Double?        = nil
    var workDatetimeFrom: Int64?     = nil

    var latitudeTo: Double?          = nil
    var longitudeTo: Double?         = nil
    var accuracyTo: Double?          = nil
    var workDatetimeTo: Int64?       = nil

    var temperature: String?          = nil
    var experiencePlaceNote1: String? = nil
    var operatingMode: String         = String()

    override init() {
        super.init()
        recordChanged()
    }

    override class var databaseTableName: String { "CommonWorkResult" }
    
    override class var primaryKey: String { "companyCd,userId,allocationNo,allocationRowNo" }

    required init(row: Row) {
        companyCd = row[CommonWorkResultColumns.companyCd]
        userId = row[CommonWorkResultColumns.userId]
        allocationNo = row[CommonWorkResultColumns.allocationNo]
        allocationRowNo = row[CommonWorkResultColumns.allocationRowNo]
        workCd = row[CommonWorkResultColumns.workCd]
        status = row[CommonWorkResultColumns.status]
        delayStatus = row[CommonWorkResultColumns.delayStatus]
        delayReasonCd = row[CommonWorkResultColumns.delayReasonCd]
        misdeliveryStatus = row[CommonWorkResultColumns.misdeliveryStatus]
        orgAllocationRowNo = row[CommonWorkResultColumns.orgAllocationRowNo]
        
        latitudeFrom = row[CommonWorkResultColumns.latitudeFrom]
        longitudeFrom = row[CommonWorkResultColumns.longitudeFrom]
        accuracyFrom = row[CommonWorkResultColumns.accuracyFrom]
        workDatetimeFrom = row[CommonWorkResultColumns.workDatetimeFrom]
        
        latitudeTo = row[CommonWorkResultColumns.latitudeTo]
        longitudeTo = row[CommonWorkResultColumns.longitudeTo]
        accuracyTo = row[CommonWorkResultColumns.accuracyTo]
        workDatetimeTo = row[CommonWorkResultColumns.workDatetimeTo]
        
        temperature = row[CommonWorkResultColumns.temperature]
        experiencePlaceNote1 = row[CommonWorkResultColumns.experiencePlaceNote1]
        super.init(row: row)
    }
    
    init(userInfo: UserInfo, binDetail: BinDetail) {
        companyCd = userInfo.companyCd
        userId = userInfo.userId
        allocationNo = binDetail.allocationNo
        allocationRowNo = binDetail.allocationRowNo
        workCd = binDetail.workCd
        status = binDetail.status
        delayStatus = binDetail.delayStatus
        delayReasonCd = binDetail.delayReasonCd
        misdeliveryStatus = binDetail.misdeliveryStatus
        
        orgAllocationRowNo = binDetail.origAllocationRowNo
        
        latitudeFrom = binDetail.startLatitude
        longitudeFrom = binDetail.startLongitude
        accuracyFrom = binDetail.startAccuracy
        workDatetimeFrom = binDetail.startDate

        latitudeTo = binDetail.endLatitude
        longitudeTo = binDetail.endLongitude
        accuracyTo = binDetail.endAccuracy
        workDatetimeTo = binDetail.endDate

        temperature = binDetail.temperature
        experiencePlaceNote1 = binDetail.experiencePlaceNote1
        
        operatingMode = binDetail.autoModeFlag ? "autoMode" : String()

        super.init()
        self.recordChanged()
    }

    override func encode(to container: inout PersistenceContainer) {
        container[CommonWorkResultColumns.companyCd] = companyCd
        container[CommonWorkResultColumns.userId] = userId
        container[CommonWorkResultColumns.allocationNo] = allocationNo
        container[CommonWorkResultColumns.allocationRowNo] = allocationRowNo
        container[CommonWorkResultColumns.workCd] = workCd
        container[CommonWorkResultColumns.status] = status
        container[CommonWorkResultColumns.delayStatus] = delayStatus
        container[CommonWorkResultColumns.delayReasonCd] = delayReasonCd
        container[CommonWorkResultColumns.misdeliveryStatus] = misdeliveryStatus
        container[CommonWorkResultColumns.orgAllocationRowNo] = orgAllocationRowNo

        container[CommonWorkResultColumns.latitudeFrom] = latitudeFrom
        container[CommonWorkResultColumns.longitudeFrom] = longitudeFrom
        container[CommonWorkResultColumns.accuracyFrom] = accuracyFrom
        container[CommonWorkResultColumns.workDatetimeFrom] = workDatetimeFrom

        container[CommonWorkResultColumns.latitudeTo] = latitudeTo
        container[CommonWorkResultColumns.longitudeTo] = longitudeTo
        container[CommonWorkResultColumns.accuracyTo] = accuracyTo
        container[CommonWorkResultColumns.workDatetimeTo] = workDatetimeTo

        container[CommonWorkResultColumns.temperature] = temperature
        container[CommonWorkResultColumns.experiencePlaceNote1] = experiencePlaceNote1
        
        super.encode(to: &container)
    }
    
    func startWorkIsNull() -> Bool {
        return latitudeFrom == nil && longitudeFrom == nil
        && accuracyFrom == nil && workDatetimeFrom == nil
    }
    
    func endWorkIsNull() -> Bool {
        return latitudeTo == nil && longitudeTo == nil
        && accuracyTo == nil && workDatetimeTo == nil
    }
}
